Handle `step` progressions in ForLoopsLowering.

punzki opened the pull request:
The first commit is from #2390.

I still need to add a ton more tests, but I'm sending this out for review of the rest of the logic which is mostly in `StepHandler.kt`. You can see the direction I'm going w.r.t. testing in the currently added box tests, and I'm going to add more bytecode text tests as well. I also moved the handlers into separate files since the single file was getting quite large.

This version handles more cases than the `ForLoopsLowering` in Kotlin/Native (such as the nested `step` case and combinations of `reverse`+`step`) and minimizes temporary variable use.

pyos commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/matchers/IrFunctionMatcher.kt:
```diff
@@ -97,3 +99,13 @@ internal fun IrFunctionMatcherContainer.singleArgumentExtension(
     return this
 }
+
+internal fun IrFunctionMatcherContainer.singleArgumentExtensionBySimpleType(
```
This looks suspicious, why not change `progressionClassesTypes` to contain IR types?

pyos replied:
Ah, already done, sorry.

punzki replied:
Oops, fixed.

pyos commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt:
```diff
@@ -221,6 +217,302 @@ internal class UntilHandler(private val context: CommonBackendContext, private v
+
+    override val matcher = SimpleCalleeMatcher {
+        singleArgumentExtensionBySimpleType(FqName("kotlin.ranges.step"), symbols.progressionClassesTypes)
+        parameterCount { it == 1 }
```
This is redundant.

pyos commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt:
```diff
@@ -221,6 +217,302 @@ internal class UntilHandler(private val context: CommonBackendContext, private v
+                return nestedInfo
+            }
+
+            // To reduce local variable usage, we create and use temporary variables only if necessary.
```
Does this actually have any effect? At least on the JVM, `JvmBuiltinOptimizationLowering` would remove these variables anyway.

punzki replied:
It does make a difference. `JvmBuiltinOptimizationLowering` does not:
1. Eliminate the variable or its initializer. Not sure if these get optimized out at the bytecode level later on, I don't believe so.
2. Replace gets of variables that are initialized by gets of variables.

Now these _could_ be addressed in the lowering or some other lowering (e.g., constant propagation or something similar).

pyos commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt:
```diff
@@ -221,6 +217,302 @@ internal class UntilHandler(private val context: CommonBackendContext, private v
+                        putValueArgument(1, zeroStep)
+                    }
+
+                    checkedStepVar = scope.createTemporaryVariable(checkedStepExpression, nameHint = "checkedStep")
```
The only variable that is created even if the value is constant - is this for a reason? Looks like it can be constant...

punzki replied:
Missed that, thanks!

pyos commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/handlers/UntilHandler.kt:
```diff
@@ -0,0 +1,168 @@
+
+    override val matcher = SimpleCalleeMatcher {
+        singleArgumentExtension(FqName("kotlin.ranges.until"), progressionElementTypes)
+        parameterCount { it == 1 }
```
Also redundant.

madsager commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/handlers/StepHandler.kt:
```diff
@@ -0,0 +1,290 @@
+            val stepType = data.stepType(context.irBuiltIns)
+            val stepGreaterFun = context.irBuiltIns.greaterFunByOperandType[stepType.toKotlinType()]!!
+            val zeroStep = if (data == ProgressionType.LONG_PROGRESSION) irLong(0) else irInt(0)
+            val throwIllegalStepExceptionCall = irCall(context.irBuiltIns.illegalArgumentExceptionSymbol, stepType).apply {
```
Maybe move the construction of this to the case where it is needed to not build the IR for it when it is not needed?

punzki replied:
Easy enough, converted to lambda

madsager commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/handlers/UntilHandler.kt:
```diff
@@ -0,0 +1,167 @@
+            //   // Standard form of loop over progression
+            //   var inductionVar = untilReceiverValue
+            //   val last = untilArg - 1
+            //   if (inductionVar <= last && untilFunArg != MIN_VALUE) {
```
untilFunArg -> untilArg

madsager commented on compiler/testData/codegen/bytecodeText/forLoop/primitiveProgression.kt:
```diff
@@ -1,3 +1,4 @@
+// IGNORE_BACKEND: JVM_IR
```
Could you add comments to the tests that now fail on JVM_IR, so it is easy for us to see why these fail and if that is by design?

punzki commented:
@max-kammerer Updated the PR. Had to resolve merge conflicts from @madsager 's changes (#2541, #2547, #2552, #2560).

Note that one of the commits (10ab88450216c8089670d5135aa148a25cd9cc56) split up the `ProgressionHandlers.kt` file which was getting pretty big. It might be easier to review without that commit.

punzki commented:
> Note that one of the commits ([10ab884](https://github.com/JetBrains/kotlin/commit/10ab88450216c8089670d5135aa148a25cd9cc56)) split up the `ProgressionHandlers.kt` file which was getting pretty big. It might be easier to review without that commit.

Updated the PR without the last commit to split up, to make it easier to review and manage potential conflicts. I'll split it up separately.

max-kammerer commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/ir/Ir.kt:
```diff
@@ -108,6 +105,12 @@ abstract class Symbols<out T : CommonBackendContext>(val context: T, private val
+        Name.identifier("getProgressionLastElement"),
+        NoLookupLocation.FROM_BACKEND
+    ).filter { it.containingDeclaration !is BuiltInsPackageFragment }
+        .mapNotNull { Pair(it.returnType!!, symbolTable.referenceSimpleFunction(it)) }.toMap()
```
Always not null values

max-kammerer replied:
@punzki Could you change mapNotNull to simple map?

max-kammerer commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt:
```diff
@@ -219,6 +220,272 @@ internal class UntilHandler(private val context: CommonBackendContext, private v
+            //
+            //   1..10 step 1               // Nested step is 1, argument is 1. Equivalent to `1..10`.
+            //   10 downTo 1 step 1         // Nested step is -1, argument is 1. Equivalent to `10 downTo 1`.
+            //   10 downTo 1 step 2 step 2  // Nested step is -2, argument is 2. Equivalent to `10 downTo 1 step 2`.
```
Is it possible to add next failing test (if they not present yet)
with  ```10 downTo 1 step 2 step -2```  and  ```1..10 step -2 step 2``` ?

punzki replied:
I have both already; see `compiler/testData/codegen/box/ranges/stepped/expression/downTo/legalStepThenIllegalStep.kt` and `compiler/testData/codegen/box/ranges/stepped/expression/rangeTo/illegalStepThenLegalStep.kt` for all various combinations. Is that what you're referring to?

max-kammerer replied:
Yes)

max-kammerer commented:
@punzki Maybe it worth to write some test generator similar to GenerateInRangeExpressionTestData/GeneratePrimitiveVsObjectEqualityTestData

max-kammerer commented on compiler/testData/codegen/bytecodeText/intRangeNoBoxing.kt:
```diff
@@ -3,16 +3,9 @@ fun foo() {
     for (i in range) {
     }
-
-    for (i in 1..2 step 4) {}
```
Why it's removed?

punzki replied:
I removed it because the behavior is different now with JVM IR (no calls to getFirst/Last). But since now we have the JVM_IR_TEMPLATES I'll just change the file.

max-kammerer replied:
Yes, please

punzki commented:
> @punzki Maybe it worth to write some test generator similar to GenerateInRangeExpressionTestData/GeneratePrimitiveVsObjectEqualityTestData

It's going to be tricky -- the tests aren't perfectly "symmetrical" and many of the values are hand-picked to cover the edge cases -- but I'll see what I can do. It'll take a while though.

punzki commented:
@max-kammerer I updated the PR but am still working on the test generator. In the meantime, I found the new tests caused problems with JS (non-IR). Filed this bug: https://youtrack.jetbrains.com/issue/KT-34166. For now, I disabled the affected tests for JS. I also had to fix duplicate IR element errors.

max-kammerer commented:
@punzki Thank you for a bug) Actually test generator could be postponed for another PR so this one could be merged

max-kammerer commented on compiler/testData/codegen/bytecodeText/intRangeNoBoxing.kt:
```diff
@@ -12,7 +12,19 @@ fun foo() {
+// JVM IR has an optimized handler for "step" progressions and elides the construction of the stepped progressions.
+
+// JVM_TEMPLATES
 // 1 INVOKEVIRTUAL kotlin/ranges/IntProgression.getFirst \(\)I
```
Please correct templates to use simple function name for 0 count (like in primitiveProgression.kt):
e.g. 
//0 INVOKEVIRTUAL kotlin/ranges/IntProgression.getFirst \(\)Ljava/lang/Integer; 
->
//0 getFirst

punzki replied:
It seems that the origin and purpose of the test is to make sure the return types for `getFirst/last/step` are not boxed. I'm splitting this up into 2 test files (range and progression) so that it's clearer.

max-kammerer replied:
Minor, but I would preferer
```
// 1 INVOKEVIRTUAL kotlin/ranges/IntProgression.getFirst \(\)I
// 1 getFirst
and so on
```
and same in intProgressionNoBoxing.kt

punzki replied:
@max-kammerer Updated the test

punzki commented:
@max-kammerer - Finally finished the test generator! Hope it looks good. 🙂 It actually caught some copy-paste issues and a missing test.

I also added the existing generators (`GenerateInRangeExpressionTestData` and `GeneratePrimitiveVsObjectEqualityTestData`) in `GenerateCompilerTests` so that they are invoked during test generation. From what I can tell, they used to have to be invoked manually. I couldn't find any code references to them anywhere -- maybe I missed it so please correct if I'm mistaken about them!

punzki commented:
@max-kammerer - Gentle ping... Let me know if you are able to look and have any other feedback!

max-kammerer commented:
@punzki Thank you for ping and sorry for defaly, will look in next days!

max-kammerer commented:
> I also added the existing generators (GenerateInRangeExpressionTestData and GeneratePrimitiveVsObjectEqualityTestData) in GenerateCompilerTests so that they are invoked during test generation. From what I can tell, they used to have to be invoked manually. I couldn't find any code references to them anywhere -- maybe I missed it so please correct if I'm mistaken about them!

Yes, they was called manually. Thank for change!

max-kammerer commented on compiler/tests/org/jetbrains/kotlin/generators/tests/GenerateInRangeExpressionTestData.kt:
```diff
@@ -28,17 +17,18 @@ object GenerateInRangeExpressionTestData {
     }
 
     private fun PrintWriter.generateTestCaseBody(header: String, rangeExpressions: List<String>, elementExpressions: List<String>) {
+        println("// KJS_WITH_FULL_RUNTIME")
```
Why this change is required? Are affected tests was broken?

max-kammerer replied:
Or this line was added manually for non affected?

punzki replied:
It was added manually before for the "in range" tests, instead of through this test generator. It was only added for the tests that needed it; there are 6 affected tests. Now we want this generator to be called during test generation. It was cleaner to add the directive for all 14 tests generated by this generator—whether it was needed or not for the test—rather than adding it only for the 6 tests that needed it.

max-kammerer commented:
@punzki Is it possible to merge tests in more bigger pieces to decrease test count?

max-kammerer commented:
Please also note that  after #2657 merge there is a problem with test
```
Caused by: java.lang.AssertionError: Unexpected IR element found during code generation. Either code generation for it is not implemented, or it should have been lowered: STRING_CONCATENATION type=kotlin.String
```

udalov commented:
In addition to Mike's question above, I wonder if it's possible to minimize the total amount of test code in the PR? Specifically, can we reduce the number of cases tested for different combinations of steps in succession, and in combination with `reversed`? Also, are all tests equally relevant both for expression and literal forms?

My main concern is the execution time of box tests, both locally and on the build server. It's probably not a big deal if we do need that many tests, but it'd be nice if they didn't take major time on each run of all black box tests.

punzki commented:
> @punzki Is it possible to merge tests in more bigger pieces to decrease test count?

Do you mean something like combining all the "illegal step" test files (there are currently 5) into a single test file? And maybe similarly for nested, step+reversed, etc? Would that make a 

> In addition to Mike's question above, I wonder if it's possible to minimize the total amount of test code in the PR? Specifically, can we reduce the number of cases tested for different combinations of steps in succession, and in combination with `reversed`? Also, are all tests equally relevant both for expression and literal forms?

There was some bug mentioned in the native `ForLoopsLowering` (now deleted) which described a bug in handling nested step loops, which is why I added those tests. I combined those with `reversed` to make sure we handle the calculations correctly. The common lowering now handles those nested-call progressions, and the native version did not.

The handling of `step` is different when the underlying progression is a literal vs some other expression, because the direction of the underlying progression is either known (for literal) or unknown. The certainty (or lack of) in direction propagates across additional `step` and `reverse` calls. So the tests are relevant.

There is certainly some redundant coverage across literal and expression forms, even across `rangeTo/..` and `until`, but I am not comfortable picking and choosing the minimum number of tests to ensure enough coverage (also I don't think I'm smart enough to choose wisely :slightly_smiling_face: ). In addition, there can be changes in the future in which the reduced coverage can come back and bite us.

> My main concern is the execution time of box tests, both locally and on the build server. It's probably not a big deal if we do need that many tests, but it'd be nice if they didn't take major time on each run of all black box tests.

I experimented with taking out some tests and ran all the codegen box tests (IR and non-IR), and measured the test time only:

- Baseline, no changes: 9757 tests, took 2m 50s
- Without "expression" tests: 9563 tests, took 2m 47s (2% fewer tests, 1.8% faster)
- Without all "stepped" ("expression" and "literal") tests: 9359 tests, took 2m 45s (4% fewer tests, 3% faster)

Do you think the difference is significant? The tests are running in parallel by test class (i.e., test data directories). I can continue to experiment in consolidating the tests (as Mike suggested), or even splitting up the tests into more directories.

punzki commented:
> Please also note that after #2657 merge there is a problem with tests in master
> 
> ```
> Caused by: java.lang.AssertionError: Unexpected IR element found during code generation. Either code generation for it is not implemented, or it should have been lowered: STRING_CONCATENATION type=kotlin.String
> ```

Fixed. I moved ForLoopsLowering before JvmStringConcatenationLowering in the JVM phases. ForLoopsLowering may produce IrStringConcatenations which should be
eliminated before codegen.

udalov commented:
Fair enough, thanks for the explanation! Let's keep the tests, but it'd be awesome if some of them can be merged like you suggest.

One other thing: consider ignoring the directories with generated tests for `AbstractLightAnalysisModeTest` in `GenerateCompilerTests`. These tests check the light analysis mode (where method bodies are not generated) on different kinds of declarations, but they won't do anything useful here because each test only consists of the single function `box`.

punzki commented:
> Fair enough, thanks for the explanation! Let's keep the tests, but it'd be awesome if some of them can be merged like you suggest.

Hmm, the generator I wrote doesn't support multiple tests per file. I could refactor it but I think it would be messy. It'd be easier to understand and debug tests in the future if there's one per file. Hope it's ok that I just left it as-is.

> One other thing: consider ignoring the directories with generated tests for `AbstractLightAnalysisModeTest` in `GenerateCompilerTests`. These tests check the light analysis mode (where method bodies are not generated) on different kinds of declarations, but they won't do anything useful here because each test only consists of the single function `box`.

Didn't know that; thanks for the info. Done!

max-kammerer commented:
@punzki Great work! Many thanks!

max-kammerer commented:
@punzki Could you take a look: there is a problem with test regeneration on Windows because of another line separator - all generated files are marked as changed cause of it. Problem is only with new generator

punzki commented:
> @punzki Could you take a look: there is a problem with test regeneration on Windows because of another line separator - all generated files are marked as changed cause of it. Problem is only with new generator

Can you describe the issue further? Do you have a TeamCity log link, or something like that? I do not have a Windows machine easily available. Also I don't really understand what the issue is...

max-kammerer commented:
@punzki Actually it's reproduced on attempt to run test generator on Windows machine (this task is not executed on Teamcity). Actually i'm using Linux too: so look like println insert \n\r on Windows that affects all generated testdata for new stepped tests
